cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -270,11 +315,55 @@ public Void call(DatasetContext context) throws Exception {
               return null;
             }
           });
+        } catch (TransactionFailureException e) {
+          readyJobsIter.remove();
+          long current = System.currentTimeMillis();
+          long nextRetryWaitMillis = retryFailedJobStartegy.nextRetry(1, current);
+          if (nextRetryWaitMillis < 0) {
+            LOG.warn("Failed to run program {} in schedule {}. Skip running this program.",
+                     job.getSchedule().getProgramId(), job.getSchedule().getName(), e);
+            continue;
+          }
+          LOG.warn("Failed to run program {} in schedule {}. Will retry in {}ms.",
+                   job.getSchedule().getProgramId(), job.getSchedule().getName(), nextRetryWaitMillis, e);
+          // add the failed job to the queue to wait for retry
+          failedJobs.add(new FailedJob(job, current, current + nextRetryWaitMillis));
+        }
+      }
+    }
 
+    private void rerunFailedJobs() {
+      final Iterator<FailedJob> failedJobsIter = failedJobs.iterator();
+      while (failedJobsIter.hasNext() && !stopping) {
+        final FailedJob failedJob = failedJobsIter.next();
+        long current = System.currentTimeMillis();
+        if (failedJob.getNextRetryTime() > current) {
+          continue;
+        }
+        try {
+          Transactions.execute(transactional, new TxCallable<Void>() {
+            @Override
+            public Void call(DatasetContext context) throws Exception {
+              if (runReadyJob(failedJob.getJob())) {
+                failedJobsIter.remove();
+              }
+              return null;
+            }
+          });
         } catch (TransactionFailureException e) {
-          LOG.warn("Failed to run program {} in schedule {}. Skip running this program.",
-                   job.getSchedule().getProgramId(), job.getSchedule().getName(), e);
-          // don't delete the job or remove it from 'readyJobs', as it will be retried in some later iteration
```

Comment:
can we ignore all this other stuff and just treat a failure as a failure? In other words, if it fails, remove it from the queue. Wasn't that the behavior before 4.3.0?

In general, if we are retrying, we need to know for sure that retrying is actually helpful. Otherwise we end up in the current situation, where it's worse than not retrying at all.